tests/pipeline_test.py

```diff
@@ -1208,3 +1208,23 @@ def process(
 def test_html_repr() -> None:
     pipeline = sl.Pipeline([make_int], params={float: 5.0})
     assert isinstance(pipeline._repr_html_(), str)
+
+
+def test_does_not_allow_type_argument_outside_of_constraints() -> None:
+    T = TypeVar('T', int, float)
+
+    @dataclass
+    class MyType(Generic[T]):
+        value: T
+
+    def process(value: T) -> MyType[T]:
+        return MyType(value)
+
+    pipeline = sl.Pipeline((process,))
+    pipeline[str] = 'abc'
+    pipeline[int] = 123
+
+    pipeline.get(MyType[int])
+
+    with pytest.raises(sl.handler.UnsatisfiedRequirement):
```

Comment:
Is this the correct exception type?

Reply:
I think so, we're requiring a `MyType[str]` but there is no provider that defines that quantity.

What exception type would you expect there?

Reply:
My point is that `MyType[str]` is invalid, whether a provider exists or not.

Reply:
Sure, and in general we check this before constructing the graph. But I still think we should raise a different exception type?